File: app/src/main/java/com/example/weatherapp/WeatherIconMapper.java
package com.example.weatherapp;

public final class WeatherIconMapper {

    public static String describe(Integer code) {
        if (code == null) return "—";
        int c = code;
        if (c == 0) return "Cerah";
        if (c == 1 || c == 2 || c == 3) return "Cerah berawan";
        if (c == 45 || c == 48) return "Berkabut";
        if (c == 51 || c == 53 || c == 55) return "Gerimis";
        if (c == 56 || c == 57) return "Gerimis beku";
        if (c == 61 || c == 63 || c == 65) return "Hujan";
        if (c == 66 || c == 67) return "Hujan beku";
        if (c == 71 || c == 73 || c == 75) return "Salju";
        if (c == 80 || c == 81 || c == 82) return "Hujan deras";
        if (c == 95) return "Badai petir";
        if (c == 96 || c == 99) return "Badai petir (es)";
        return "Cuaca";
    }

    private WeatherIconMapper() {}
}

File: app/src/main/java/com/example/weatherapp/WeatherIconMapper.kt
package com.example.weatherapp

object WeatherIconMapper {

    private fun isNight(localHour: Int) = localHour < 6 || localHour >= 18

    fun iconRes(code: Int, localHour: Int): Int {
        val night = isNight(localHour)
        return when {
            code == 0 -> if (night) R.drawable.icon02n else R.drawable.icon02d
            code in 1..3 || code in 45..48 -> if (night) R.drawable.icon03n else R.drawable.icon03d
            code in 51..67 || code in 80..82 || code in 95..99 -> R.drawable.icon09n
            else -> if (night) R.drawable.icon03n else R.drawable.icon03d
        }
    }

    fun description(code: Int): String = when (code) {
        0 -> "Cerah"
        1 -> "Cerah sebagian"
        2 -> "Berawan sebagian"
        3 -> "Berawan"
        in 45..48 -> "Berkabut"
        in 51..57 -> "Gerimis"
        in 61..67 -> "Hujan"
        in 71..77 -> "Salju"
        in 80..82 -> "Hujan lokal"
        in 95..99 -> "Badai petir"
        else -> "Cuaca tidak diketahui"
    }
}
